runtime/interpreter/value.go

```diff
@@ -2946,6 +2968,100 @@ func (v *ArrayValue) Reverse(
 	)
 }
 
+func (v *ArrayValue) Filter(
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	procedure FunctionValue,
+) Value {
+	filteredValuesCount := 0
+
+	invocation := NewInvocation(
+		interpreter,
+		nil,
+		nil,
+		[]Value{}, // Set later during invocation.
+		[]sema.Type{v.semaType.ElementType(false)},
+		nil,
+		locationRange,
+	)
+	iterationInvocation := func(arrayElement Value) Invocation {
+		invocation.Arguments = []Value{arrayElement}
+		return invocation
+	}
+
+	err := v.array.Iterate(
```

Comment:
Why is it needed to iterate through the array twice? once here (seems only to find the count? can it be removed), and again in the iterator of the new array (make sense)

Reply:
Yeah, this iteration should be removed.

`NewArrayValueWithIterator` requires an element count, but it is only used for memory metering (see `newArrayValueFromConstructor`). Maybe its `count` parameter should be renamed to something like `countOverestimate`, and in this function we simply overestimate to the count of the original array (worst case all elements are included in the final array)

Reply:
Done, I didn't know we could overestimate. Good to know.